prefetch.bs

```diff
@@ -445,6 +454,46 @@ The <dfn>list of sufficiently strict speculative navigation referrer policies</d
     </div>
 </div>
 
+<h2 id="cookies">Cookies</h2>
+
+[[COOKIES]] defines "cookies" which can be set using the <a http-header spec="COOKIES" lt="Set-Cookie">`` `Set-Cookie` ``</a> response header field. Because [=uncredentialed prefetch=] forces a separate [=network partition key=] to be used, it's necessary to copy these cookies into the ordinary partition as though they had been [=receive a cookie|received=] at the time of navigation.
+
+<div algorithm>
+  To <dfn>copy prefetch cookies</dfn> given a [=network partition key=] |isolatedPartitionKey| and an [=environment=] |environment|, perform the following steps.
+
+  <div class="note">
+    Though formally there is only one cookie store in [[COOKIES]], some browsers partition cookie stores so as to separate, for example, cookies with the same domain when loaded with different top-level sites. See <a href="https://github.com/privacycg/storage-partitioning">Client-Side Storage Partitioning (Privacy CG)</a>.
+  </div>
+
+  1. Let |isolatedCookieStore| be the cookie store associated with |isolatedPartitionKey|.
+  1. For each <a spec="COOKIES">cookie</a> |cookie| in |isolatedCookieStore|.
+      1. Remove |cookie| from |isolatedCookieStore|.
+      1. A user agent may ignore a cookie in its entirety. If so, continue.
+
+          <div class="note">This is consistent with [[COOKIES]] expressly permitting this when [=receiving a cookie=].</div>
+      1. Let |topLevelSite| be null.
+      1. If |environment|'s [=environment/target browsing context=] is a [=top-level browsing context=]:
```

Comment:
I'm probably missing something, but why is this special-cased? Couldn't we get the information directly from _environment_ like the "Otherwise:" case does?

Reply:
Because the top-level site at the time the cookie is assigned would have been determined not by wherever the navigation ended up (which is what environment reflects at this time), but by the URL that was being fetched then. The two are not necessarily equal, because if we're following redirects (the spec assumes we do), then we might have picked up cookies from those origins too (e.g., we might have followed a redirect from example.com to www.example.com, and if the former set a cookie it should be copied faithfully).

Reply:
So to confirm I understand, this means what we have here only works for prefetches (where child browsing contexts are not fetched before the redirect kicks in), or prerenders where we avoid loading child browsing contexts until activation. If we ever used this for a version of cross-origin prerendering which loaded child browsing contexts before activation, then we'd need to do a similar dance for establishing the right top-level site for their cookies. Right?

This feels a bit delicate and maybe worth explaining in a couple notes. Here's my attempt:

> When copying the top-level browsing context's cookies, we use the cookie's domain to establish _topLevelSite_ and thus _destinationPartitionKey_. The cookie might have been delivered at any point in the redirect chain while fetching the top-level document, so its domain is not necessarily equal to the site corresponding to _environment_'s top-level origin.

> When copying nested browsing contexts' cookies, we know that _environment_'s top-level origin, which reflects the final destination of the top-level browsing context, is sufficient to establish _topLevelSite_ and thus _destinationPartitionKey_. This is because we only fetch nested browsing context documents after any top-level document redirects have ceased.

Reply:
Well, close. Since this is for prefetch, either the whole prefetch relates to a top-level navigation (in which case every fetch, redirect or no, is top-level and so the top-level site matches that of the cookie's domain) or the whole prefetch occurs in some nested browsing context (in which case in which case the top-level site is that of the top-level browsing context).

Because subframes of the prefetched resource don't exist until document parsing (and nothing currently parses the document), there are no cookies thereto appertaining.

But yes the logic for merging cookies for prerendering including subframes (in fact, including any subresources) would need more verbiage to do this correctly. Much of this complexity just arises from the fact that these abstractions don't really exist in the cookie spec and only kinda exist in Fetch.

Added two notes (one rather long) in the hopes of clarifying.